### PR TITLE
Fix and enhance recursive pruning

### PR DESCRIPTION
**Pull request by PhilZ-cwm6**
Currently, `-R` option removes all child dataset snapshots, including any cloned dataset outside of the specified pool descendants. The problem is that the actual code, using `-l` or `-n` doesn't show these destroyed clones !

It is more secure to rely on `-n`, `-r` and `-R` options from the `zfs destroy`command rather than adding bugs by customizing them.

The current commit:
- adds `-r` and `-R` support as they are expected to work with `zfs destroy`. When deleteing snapshots recursively, we no longer need to provide every single child snapshot to the the `zfs destroy`command. Instead, only the specified pool/dataset is passed to the `zfs destroy` command and we rely on it to destroy the related child snapshots
- Now, we can prune snapshots in a non recursive way, which was not possible in original script
- adds `-n` dry-run option directly into `zfs destroy`, so that the actual snapshots that would be destroyed are always properly displayed to the user in a dry-run
- adds support to the `-v` option natively in `zfs destroy`:  during a recursive delete, the removed child datasets are properly displayed
- rename the previous `-v` option to `-d` (debug) to show more debug messages like the skipped datasets (unchanged)
- add `-e` option to abort on a failed `zfs destroy` command
- fix exit code could be wrongly set to other value than 0 on final exit

### CODE REVIEW

**Comment by PhilZ-cwm6** [comment]
I loved your script, thank you a lot. However, I fixed a few potential issues, and mainly, I added a non recursive delete + enhanced the recursive, fixed potential bugs that could result in data loss using -R and added more native zfs destroy command processing

File: zfs-prune-snapshots [hunk 1]
```diff
@@ -1,12 +1,12 @@
-#!/usr/bin/env bash
+#!/bin/bash
```
Comment (HRio) [inline zfs-prune-snapshots]: `bash` in not in `/bin` on all systems with ZFS

File: zfs-prune-snapshots [hunk 2]
```diff
@@ -353,15 +382,20 @@ for line in "${lines[@]}"; do
...
 		echo -n '[dry-run] '
 	fi
 
 	echo "[$i/$numsnapshots] removing $snapshot: $ht old ($hu)"
 
-	if ! $dryrun; then
-		zfs destroy "${destroyargs[@]}" "$snapshot" || code=1
-	fi
+	zfs destroy "${destroyargs[@]}" "$snapshot"
+    code=$?
```
Comment (HRio) [inline zfs-prune-snapshots]: wrong indentation here and on line 385?

File: zfs-prune-snapshots (hunk 1 above)
Comment (PhilZ-cwm6) [reply zfs-prune-snapshots]: Ah, sorry, I just reverted it, I use teh script under FreeBSD

File: zfs-prune-snapshots (hunk 2 above)
Comment (PhilZ-cwm6) [reply zfs-prune-snapshots]: I fixed the identation using tabs instead of spaces

File: zfs-prune-snapshots (hunk 1 above)
Comment (PhilZ-cwm6) [reply zfs-prune-snapshots]: I did not aim to deeply modify your script, but using `#!/usr/bin/env` can be useful under perl or python to specify specific versions for example, but for bash, I am not sure it is a good way + the potential security risks depending on the plateforms / use. But it is another debate like using `getopts` instead of a manual parser with input sanitization. Anyway, hope you merge the fixes because the recursiveness wasn't really ideal: no way to not prune recursively, and the unexpected deletes of cloned snapshots using `-R`

**Comment by PhilZ-cwm6** [comment]
i can squash in a single commit once done

**Comment by bahamas10** [comment]
Hello! thank you so much for this PR - so far I've just read the description and all of it seems good! i'm about to review the code and test it :)

**Comment by bahamas10** [comment]
Hello! thank you so much for this PR - so far I've just read the description and all of it seems good! i'm about to review the code and test it :)

File: zfs-prune-snapshots [hunk 3]
```diff
@@ -53,23 +53,28 @@ timespec
...
-		   without deleting (like dry-run mode with machine-parseable
-		   output)
-    -p <prefix>    snapshot prefix string to match
-    -s <suffix>    snapshot suffix string to match
-    -i             invert matching of prefix and suffix
-    -q             quiet, do not printout removed snapshots
-    -R             recursively delete, pass '-R' directly to 'zfs destroy'
-    -v             increase verbosity
-    -V             print the version number and exit
+    -h             Print this message and exit
```
Comment (bahamas10) [inline zfs-prune-snapshots]: I see that you've added lines for `-d` and `-r`, etc. but you've also changed the beginning of each line to uppercase.  I'm happy to have a discussion on a PR for changing the casing, but as it stands right now it's half uppercase and half lowercase.... you can see this when running with `-h`.

I'd prefer you just leave the casing as-is and if you'd like to make changes to it we can handle it in a separate PR.

Personally, I have no strong feelings either way when it comes to uppercase vs lowercase in code... i tend to just not use capitals when i type out of laziness ha

Basically, i'd just prefer to keep style changes separate from function changes.

File: zfs-prune-snapshots [hunk 4]
```diff
@@ -184,8 +192,15 @@ while getopts 'hniqlRp:s:vV' option; do
 		p) prefix=$OPTARG;;
 		s) suffix=$OPTARG;;
 		q) quiet=true; exec 1>/dev/null;;
-		R) recursive=true;;
-		v) ((verbosity++));;
+		e) error_exit=true;;
+		r) recursive=true
```
Comment (bahamas10) [inline zfs-prune-snapshots]: I see that specifying `-rR` or `-Rr` will result in the last option that is set "winning"... would it make more sense to just error if the user specifies both `-r` and `-R`?

For example, have a block of code below the argument parsing like:

``` bash
if $recursive && $recursive_extended; then
  echo 'cannot specify -r and -R together' >&2
  exit 1
fi
```

File: zfs-prune-snapshots [hunk 5]
```diff
@@ -353,14 +381,19 @@ for line in "${lines[@]}"; do
 	ht=$(human-time "$delta")
 	hu=$(human-size "$used")
 
+	echo ""
```
Comment (bahamas10) [inline zfs-prune-snapshots]: i don't prefer the extra spacing here, please remove.

like i commented above, since this is a style-only change I'm happy to discuss it in a change where the function of the code isn't being modified.

File: zfs-prune-snapshots [hunk 6]
```diff
@@ -353,14 +381,19 @@ for line in "${lines[@]}"; do
...
 	fi
 
 	echo "[$i/$numsnapshots] removing $snapshot: $ht old ($hu)"
 
-	if ! $dryrun; then
-		zfs destroy "${destroyargs[@]}" "$snapshot" || code=1
+	zfs destroy "${destroyargs[@]}" "$snapshot"
+	code=$?
+
+	if $error_exit && [ $code -ne 0 ]; then
```
Comment (bahamas10) [inline zfs-prune-snapshots]: replace this with a bashism check:

``` bash
if $error_exit && ((code != 0)); then
```

File: zfs-prune-snapshots [hunk 7]
```diff
@@ -6,7 +6,7 @@
 # Date: November 20, 2015
 # License: MIT
 
-VERSION='v1.5.0'
+VERSION='v1.5.1'
```
Comment (bahamas10) [inline zfs-prune-snapshots]: Please bump the minor version here, as this PR changes the usage and arguments:

``` bash
VERSION='v1.6.0'
```

**Review by bahamas10 — CHANGES_REQUESTED** [review]
First of all, thank you so much for the change! I can tell you were very thoughtful with the features you implemented here and I think all of them are *REALLY* good!

I'm happy to see more of the functionality leaning on what the `zfs` command can do, as opposed to us trying our best to recreate what it does.

Overall i'm 👍 on the change! Just a couple of style and nit changes for you to review.

File: zfs-prune-snapshots (hunk 3 above)
Comment (PhilZ-cwm6) [reply zfs-prune-snapshots]: sure, I just pushed it quickly, I will revert to lower casing

File: zfs-prune-snapshots (hunk 7 above)
Comment (PhilZ-cwm6) [reply zfs-prune-snapshots]: ok, I will commit asap, best regards

Comment (PhilZ-cwm6) [reply zfs-prune-snapshots]: I even would reverse the version increment, and let you push a separate commit with the more eventual changes and the proper version upgrade you find useful

**Comment by PhilZ-cwm6** [comment]
I pushed the requested changes and fixed a few space/tabs I missed previously. Let me know if more formatting should be fixed

However, after more testing in real scenarios, I discovered a bug/missing feature in zfs and opened a bug/feature request on github:
https://github.com/openzfs/zfs/issues/13806

I don't expect any fix soon as they are busy on more important stuff

Maybe, for safety, you can go with any of these options, or better a combination of them:
1- add  a warning when using `-R -l` that using `-l` doesn't list dependant clones, and that if `-R` is planned, it is better to use `-n` instead of -`l`
2- when using `-Rn`, actually run a `-rn` to properly list dependant clones. This really wired, but there is no other way to list and expose the dependant clones before deleting them
3- Advise users to always start by running `-rnv` before ever running `-R`

I propose:
- the current commit gets merged as it enhances/fixes current code behaviour
- you implement the proposed fixes above as you wish in later commits
- once zfs upstream fixes these confusing outputs, you can revert that last workaround commit to zfs bug

I let you decide and modify the code as you wish. I'll give a feedback once you're done
Best regards

**Comment by PhilZ-cwm6** [comment]
I pushed the requested changes and fixed a few space/tabs I missed previously. Let me know if more formatting should be fixed

I also squashed all commits in a single one

However, after more testing in real scenarios, I discovered a bug/missing feature in zfs and opened a bug/feature request on github:
https://github.com/openzfs/zfs/issues/13806

I don't expect any fix soon as they are busy on more important stuff

Maybe, for safety, you can go with any of these options, or better a combination of them:
1- add a warning when using -R -l that using -l doesn't list dependant clones, and that if -R is planned, it is better to use -n instead of -l
2- when user asks for -Rn, actually the code should run a -rn to properly list dependant clones. This is really wired, but there is no other way to list and expose the dependant clones before deleting them
3- Advise users to always start by running -rnv before ever running -R

I propose:

the current commit gets merged as it enhances/fixes current code behaviour
you implement the proposed workarounds above as you wish in later commits
once zfs upstream fixes these inconsistant command outputs, you can revert the last workaround commit
I let you decide and modify the code as you wish. I'll give a feedback once you're done
Best regards

**Comment by PhilZ-cwm6** [comment]
I added a commit to address the currently reported limitation
Tested using the example I listed in https://github.com/openzfs/zfs/issues/13806 and it works now as expected

Sadly, there is no other way to catch cloned systems before arasing then using `-R` unless it is fixed upstream by implementing the options in `zfs list -R` or at least fix `zfs destroy -Rnv`

This was actually really interesting to discover this limitation in zfs !

**Comment by bahamas10** [comment]
so i'm having trouble while testing this, and i don't quite understand the intent.

for example, on my machine:

```
$ zfs list -r -t snapshot goliath/backup-tools
NAME                                                USED  AVAIL  REFER  MOUNTPOINT
goliath/backup-tools@automated_monthly_1656670020      0      -  62.8K  -
goliath/backup-tools@automated_monthly_1659348420      0      -  62.8K  -
goliath/backup-tools@automated_monthly_1662026820      0      -  62.8K  -
goliath/backup-tools@automated_monthly_1664614980  55.8K      -   100K  -
goliath/backup-tools@automated_monthly_1667293380  46.5K      -   185K  -
goliath/backup-tools@automated_monthly_1669888980  79.1K      -   261K  -
goliath/backup-tools@automated_monthly_1672567380  50.0K      -   340K  -
goliath/backup-tools@automated_yearly_1672567440   51.2K      -   341K  -
goliath/backup-tools@zincrsend_1672999800          51.2K      -   358K  -
goliath/backup-tools@zincrsend_1673086200          54.7K      -   362K  -
...
```

fyi: this dataset does NOT have any children.

Using your branch, i can't get this to work using the normal usage:

```
$ ./zfs-prune-snapshots -n 1M goliath/backup-tools
cannot open 'goliath/backup-tools': missing '@' delimiter in snapshot name
found 0 snapshots (0 B) on pools: goliath/backup-tools
removed 0 snapshots (0 B)
```

on the latest master branch, this works as expected.

```
$ ./zfs-prune-snapshots -n 1M goliath/backup-tools
found 9 snapshots (333 KB) on pools: goliath/backup-tools
[dry-run] [1/9] removing goliath/backup-tools@automated_monthly_1656670020: 7 months old (0 B)
[dry-run] [2/9] removing goliath/backup-tools@automated_monthly_1659348420: 6 months old (0 B)
[dry-run] [3/9] removing goliath/backup-tools@automated_monthly_1662026820: 5 months old (0 B)
[dry-run] [4/9] removing goliath/backup-tools@automated_monthly_1664614980: 4 months old (55 KB)
[dry-run] [5/9] removing goliath/backup-tools@automated_monthly_1667293380: 3 months old (46 KB)
[dry-run] [6/9] removing goliath/backup-tools@automated_monthly_1669888980: 2 months old (79 KB)
[dry-run] [7/9] removing goliath/backup-tools@automated_monthly_1672567380: 1 month old (50 KB)
[dry-run] [8/9] removing goliath/backup-tools@automated_yearly_1672567440: 1 month old (51 KB)
[dry-run] [9/9] removing goliath/backup-tools@zincrsend_1672999800: 1 month old (51 KB)
removed 9 snapshots (333 KB)
```

it seems like the logic to pass snapshots to `zfs destroy` is failing here on your branch.

---

## Questions

- does `zfs-prune-snapshots` need `-r`? like the latest master branch, it seems that `-r` should be assumed, and then `-R` can be optional to further enhance what is done by this script.
- how should we continue? if i merge this as-is i'll have to make changes to it so it works with basic usage.  alternatively, i can commit small features one-at-a-time (or you can submit smaller PRs if you'd like)....  features like:
    - `-e` for error-exit is a good stand-alone feature

**Comment by PhilZ-cwm6** [comment]
First, -r and -R are now optional, they are not assumed by default
I cannot reproduce your issue. 
```
zfs-prune-snapshots.sh -n 1M main_data/tank

[dry-run] [1/284] removing main_data/tank@auto-H6-2021-10-21_15-53: 1 year old (144 KB)

[dry-run] [2/284] removing main_data/tank@auto-H6-2021-11-25_00-00: 1 year old (0 B)

[dry-run] [3/284] removing main_data/tank@auto-H6-2021-12-24_00-00: 1 year old (0 B)

[dry-run] [4/284] removing main_data/tank@auto-H6-2021-12-26_00-00: 1 year old (0 B)

[dry-run] [5/284] removing main_data/tank@auto-H6-2021-12-30_00-00: 1 year old (0 B)

[dry-run] [6/284] removing main_data/tank@auto-H6-2021-12-31_00-00: 1 year old (0 B)
...
```
Currently, I have no more time to look at this as it works as intended in my testing
If any bugs are mentioned, I would look at it

**Comment by Fire-Dragon-DoL** [comment]
for what is worth, I did incur dataloss with the "recursive snapshot" existing option (not with this PR)

**Comment by bahamas10** [comment]
ok, i'm back again looking into the issue here.  i like this PR and want to merge it but it is currently broken for me.  i've banged my head against it for about 2 hours until i FINALLY got to the bottom of it.

This PR makes it so when i run:

```
$ ./zfs-prune-snapshots -n 1s rpool/dave-test
cannot open 'rpool/dave-test': missing '@' delimiter in snapshot name
found 0 snapshots (0 B) on pools: rpool/dave-test
removed 0 snapshots (0 B)
```

I get an error from `zfs list`.... the underlying command that fails is:

```
$ zfs list -t snapshot rpool/dave-test
cannot open 'rpool/dave-test': missing '@' delimiter in snapshot name
```

This doesn't work on my machine.

AND I KNOW WHY

OpenZFS has a specific set of code in `zfs_main.c` to handle this exact scenario:

https://github.com/openzfs/zfs/blob/d0a91b9f88a47316158508bf304a61baa8c99c10/cmd/zfs/zfs_main.c#L3979-L3987

This code block is absent on the illumos version of zfs (which I run):

https://github.com/illumos/illumos-gate/blob/45744051679350ee063cdc366b66bee5223a11ea/usr/src/cmd/zfs/zfs_main.c#L3462

I'm thinking of a way to handle this exact scenario, because I agreed that mirroring the `-r` and `-R` arguments in this script to what `zfs destroy` provides is the ideal path forward.

---

@Fire-Dragon-DoL I'm sorry if this script has caused you any data-loss - that's horrible and not my goal with it (well, intentional data-loss is the goal but you get my point).  I'm hoping to figure out exactly what to do here and come up with a solution.

**Comment by bahamas10** [comment]
ok, i'm able to modify the `zfs list` to use `-r -d 1`... this makes my illumos version of `zfs` match the version that openzfs has... the problem is this branch is still not working at all how i expect.

I have the following snapshots:

```
dave - GZ:amon sunos ~/dev/new-zfs (git:recursive_fix) $ zfs list -t snapshot -r rpool/dave-test
NAME                               USED  AVAIL     REFER  MOUNTPOINT
rpool/dave-test/bar@new-snapshot     0B      -       24K  -
rpool/dave-test/foo@snapshot         0B      -       24K  -
```

And i can't for the life of me remove them with `zfs-prune-snapshots` even when setting the timespec to 1s:

```
dave - GZ:amon sunos ~/dev/new-zfs (git:recursive_fix) $ ./zfs-prune-snapshots  1s rpool/dave-test
found 0 snapshots (0 B) on pools: rpool/dave-test
removed 0 snapshots (0 B)
dave - GZ:amon sunos ~/dev/new-zfs (git:recursive_fix) $ ./zfs-prune-snapshots -r 1s rpool/dave-test
found 0 snapshots (0 B) on pools: rpool/dave-test
removed 0 snapshots (0 B)
dave - GZ:amon sunos ~/dev/new-zfs (git:recursive_fix) $ ./zfs-prune-snapshots -R 1s rpool/dave-test
found 0 snapshots (0 B) on pools: rpool/dave-test
removed 0 snapshots (0 B)
```

I know this is because the initial pass to list the snapshots before removing them is not working... i don't know how this branch is working unless there is *another* discrepancy i'm not aware of

**Comment by Fire-Dragon-DoL** [comment]
> ok, i'm back again looking into the issue here. i like this PR and want to merge it but it is currently broken for me. i've banged my head against it for about 2 hours until i FINALLY got to the bottom of it.
> 
> This PR makes it so when i run:
> 
> ```
> $ ./zfs-prune-snapshots -n 1s rpool/dave-test
> cannot open 'rpool/dave-test': missing '@' delimiter in snapshot name
> found 0 snapshots (0 B) on pools: rpool/dave-test
> removed 0 snapshots (0 B)
> ```
> 
> I get an error from `zfs list`.... the underlying command that fails is:
> 
> ```
> $ zfs list -t snapshot rpool/dave-test
> cannot open 'rpool/dave-test': missing '@' delimiter in snapshot name
> ```
> 
> This doesn't work on my machine.
> 
> AND I KNOW WHY
> 
> OpenZFS has a specific set of code in `zfs_main.c` to handle this exact scenario:
> 
> https://github.com/openzfs/zfs/blob/d0a91b9f88a47316158508bf304a61baa8c99c10/cmd/zfs/zfs_main.c#L3979-L3987
> 
> This code block is absent on the illumos version of zfs (which I run):
> 
> https://github.com/illumos/illumos-gate/blob/45744051679350ee063cdc366b66bee5223a11ea/usr/src/cmd/zfs/zfs_main.c#L3462
> 
> I'm thinking of a way to handle this exact scenario, because I agreed that mirroring the `-r` and `-R` arguments in this script to what `zfs destroy` provides is the ideal path forward.
> 
> @Fire-Dragon-DoL I'm sorry if this script has caused you any data-loss - that's horrible and not my goal with it (well, intentional data-loss is the goal but you get my point). I'm hoping to figure out exactly what to do here and come up with a solution.

No problem and thank you. This is open source software and I'm aware it comes with no warranty of any kind. I was able to "rebuild" the lost data in a way or another, it took a while but was able to restore everything.

Thank you again for your kind words ❤️
